File: 0415-add-strings/0415-add-strings.cpp
class Solution {
public:
    string addStrings(string num1, string num2) {
        string sum = "";
        int maxLength = max(num1.size(), num2.size());
        int carry = 0;
        int tempSum;
        int start1;
        int start2;
        for(int i = 0; i < maxLength; i++){
            start1 = num1.size() - i -1;
            start2 = num2.size() - i -1;
            if (start1 < 0){
                tempSum = num2[start2] + carry - '0';
            }
            else if (start2 < 0){
                tempSum = num1[start1] + carry - '0';
            }
            else{
                tempSum = num1[start1] + num2[start2] + carry - '0'*2;
            }
            if (carry > 0){
                carry = 0;
            }
            if (tempSum > 9){
                carry = tempSum / 10;
                tempSum %= 10;
            }
            sum += to_string(tempSum);
        }
        if (carry > 0){
            sum += to_string(carry);
        }
        reverse(sum.begin(), sum.end());
        return sum;  
    }
};

// do long addition
// however in c++ to a single letter in string is considered a char, and char represents ascii values
// to convert ascii to int, you would need to subtract from the ascii '0'
// intuition here is to start from the back to the front
// start1 and start2 will always point to the furtherest back value of a string regardless of length, then moves forward

File: 0415-add-strings/0415-add-strings.py
class Solution:
    def addStrings(self, num1: str, num2: str) -> str:
        sum = ""
        maxLength = max(len(num1), len(num2))
        carry = 0
        for i in range(maxLength):
            start1 = len(num1) - i -1
            start2 = len(num2) - i -1
            if (start1 < 0):
                tempSum = int(num2[start2]) + carry
            
            elif (start2 < 0):
                tempSum = int(num1[start1]) + carry
            
            else:
                tempSum =int(num1[start1]) + int(num2[start2]) + carry
            
            carry = tempSum // 10
            tempSum %= 10
            sum += str(tempSum)
        
        if (carry > 0):
            sum += str(carry)

        
        return sum[::-1]
        
# intuition is to iterate from the furtherest element, perform additions then move forward
# if can't go forward then ignore that number, or treat as 0